apps/builder/app/builder/shared/assets/use-assets.tsx

```diff
@@ -54,100 +43,81 @@ const setAsset = (asset: Asset) => {
   });
 };
 
-export type UploadData = ActionData;
+type FileData = {
+  id: string;
+  type: AssetType;
+  file: File;
+  objectURL: string;
+};
 
-const toUploadingAssetsAndFormData = (
-  type: AssetType,
-  files: File[]
-): Promise<[UploadingAssetContainer, FormData][]> => {
-  const assets: Array<Promise<[UploadingAssetContainer, FormData]>> = [];
+const getFilesData = (type: AssetType, files: File[]): FileData[] => {
+  return files.map((file) => {
+    return {
+      id: crypto.randomUUID(),
+      type,
+      file,
+      objectURL: URL.createObjectURL(file),
```

Comment:
In order to remove mem leaks from what I know you should call `revokeObjectURL()` after upload completes

Reply:
But that can cause the image blink if the uploaded Image is not loaded, not sure. Probably can be called for all uploaded images after few minutes delay.

Reply:
Anyway checked uploading ton of huge files, don't see any leak.

Reply:
Seems revokeObjectURL exists ;-)